download_data/about_sitemap/download_pdf_by_sitemap_to_loacl.py

```diff
@@ -0,0 +1,156 @@
+from lxml import etree
+import re
+from tqdm import tqdm 
+import json
+import multiprocessing
+
+
+headers = {
+    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36",
+}
+
+error_url_save_path = "./error_url.txt"
+
+def save_error_url(write_text):
+    if os.path.isfile(error_url_save_path):
+        mode = 'a'
+    else:
+        mode = 'w'
+        
+    with open(error_url_save_path, mode) as f:
+         f.write(write_text + '\n')
+
+def get_network_pdf(url, headers=headers, retries=3, backoff_factor=0.5):
+    session = requests.Session()
+    retry = Retry(total=retries, backoff_factor=backoff_factor)
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount('http://', adapter)
+    session.mount('https://', adapter)
+  
+    try:
+        response = session.get(url, headers=headers)
+        response.raise_for_status()
+    except Exception as e:
+        save_error_url(url + "\n" + str(e))
+        return None
+    
+    return response.content
+
+
+def get_sitemap(gz_url):
+    response = urllib.request.urlopen(gz_url)
+    compressed_file = io.BytesIO(response.read())
+    decompressed_file = gzip.GzipFile(fileobj=compressed_file)
+    return decompressed_file.read()
+
+
+def get_file_url_in_sitemap(sitemap_text):
+    root = etree.fromstring(sitemap_text)
+    locs = []
+    for url in root.findall('{http://www.sitemaps.org/schemas/sitemap/0.9}url'):
+        loc = url.find('{http://www.sitemaps.org/schemas/sitemap/0.9}loc').text
+        locs.append(loc)
+    return locs
+
+def get_sitemap_url_in_sitemap(sitemap_text):
+    root = etree.fromstring(sitemap_text)
+    locs = []
+    for sitemap in root.findall('{http://www.sitemaps.org/schemas/sitemap/0.9}sitemap'):
+        loc = sitemap.find('{http://www.sitemaps.org/schemas/sitemap/0.9}loc').text
+        locs.append(loc)
+    return locs
+
+
+languages = ["ar", "en", "es", "fr", "ru", "zh"]
+
+language_pattern = "|".join(languages)
+language_regex = re.compile(f"({language_pattern})\.pdf$")
+
+def match_six_countries_file_url(url_list):
+    
+    pdf_dict = {}
+
+    for i in range(len(url_list) - len(languages) + 1):
+   
+        match = True
+        for j in range(len(languages)):
+            if not language_regex.search(url_list[i+j].lower()):
+                match = False
+                break
+   
+        if match:
+            file_name = url_list[i].split("/")[-1].split("-")[0]
+            pdf_dict[file_name] = [url_list[i+j] for j in range(len(languages))]
+            
+    return pdf_dict
+
+
+def check_folder_exists(dir_path):
+    if not os.path.exists(dir_path):
+        os.makedirs(dir_path)
+
+def save_pdf_file(file_path, pdf_text):
+    if pdf_text:
+        with open(file_path, mode='wb') as f:
+             f.write(pdf_text)
+
+def save_six_countries_file(file_name, pdf_urls):
+    check_folder_exists(f"{download_pdf_file_root_path}/{file_name}")
```

Comment:
import os
import requests
from pathlib import Path


def save_six_countries_file(file_name: str, pdf_urls: list):
    # Create the directory for the PDF files
    pdf_dir = Path(download_pdf_file_root_path) / file_name
    pdf_dir.mkdir(parents=True, exist_ok=True)
    
    for pdf_url in pdf_urls:
        # Get the filename from the URL
        pdf_filename = pdf_url.split('/')[-1]
        pdf_path = pdf_dir / pdf_filename
        
        # Check if the file already exists
        if not pdf_path.exists():
            # Download the PDF file
            try:
                pdf_content = get_network_pdf(pdf_url)
            except requests.exceptions.RequestException as e:
                print(f"Error downloading {pdf_url}: {e}")
                continue
                
            # Save the PDF file
            with open(pdf_path, "wb") as f:
                f.write(pdf_content)